lnrpc/routerrpc/router_server.go

```diff
@@ -487,6 +492,156 @@ func toRPCPairData(data *routing.TimedPairResult) *PairData {
 	return &rpcData
 }
 
+// ImportMissionControl imports the state provided to our internal mission
+// control. Only entries that are fresher than our existing state will be used.
+func (s *Server) ImportMissionControl(ctx context.Context,
+	req *ImportMissionControlRequest) (*ImportMissionControlResponse,
+	error) {
+
+	if len(req.Pairs) == 0 {
+		return nil, errors.New("at least one pair required for import")
+	}
+
+	snapshot := &routing.MissionControlSnapshot{
+		Pairs: make(
+			[]routing.MissionControlPairSnapshot, len(req.Pairs),
+		),
+	}
+
+	for i, pairResult := range req.Pairs {
+		pairSnapshot, err := toPairSnapshot(pairResult)
+		if err != nil {
+			return nil, err
+		}
+
+		snapshot.Pairs[i] = *pairSnapshot
+	}
+
+	err := s.cfg.RouterBackend.MissionControl.ImportHistory(snapshot)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ImportMissionControlResponse{}, nil
+}
+
+func toPairSnapshot(pairResult *PairHistory) (*routing.MissionControlPairSnapshot,
+	error) {
+
+	from, err := route.NewVertexFromBytes(pairResult.NodeFrom)
+	if err != nil {
+		return nil, err
+	}
+
+	to, err := route.NewVertexFromBytes(pairResult.NodeTo)
+	if err != nil {
+		return nil, err
+	}
+
+	if bytes.Equal(from[:], to[:]) {
+		return nil, errors.New("peer pubkeys must differ")
+	}
+
+	failAmt, failTime, err := getPair(
+		lnwire.MilliSatoshi(pairResult.History.FailAmtMsat),
+		btcutil.Amount(pairResult.History.FailAmtSat),
+		pairResult.History.FailTime,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("pair: %v->%v invalid failure: %v",
+			from, to, err)
+	}
+
+	successAmt, successTime, err := getPair(
+		lnwire.MilliSatoshi(pairResult.History.SuccessAmtMsat),
+		btcutil.Amount(pairResult.History.SuccessAmtSat),
+		pairResult.History.SuccessTime,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("pair: %v->%v invalid success: %v",
+			from, to, err)
+	}
+
+	if successAmt == 0 && failAmt == 0 {
+		return nil, fmt.Errorf("either success of failure result " +
```

Comment:
Similarly maybe we could print the from/to nodes to make it easier to debug if there's anything wrong with the dump.